Adding support for 'minified' parameter when fetching a library's items

r0adkll opened the pull request:
<!--
For Work In Progress Pull Requests, please use the Draft PR feature,
see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.

If you do not follow this template, the PR may be closed without review.

Please ensure all checks pass.
If you are a new contributor, the workflows will need to be manually approved before they run.
-->

## Brief summary

<!-- Please provide a brief summary of what your PR attempts to achieve. -->

This PR attempts to fix the issue where the `/api/libraries/<id>/items` endpoint only returns minified library items.

## Which issue is fixed?

<!-- Which issue number does this PR fix? Ex: "Fixes #1234" -->
Fixes #2123 

## In-depth Description

<!--
Describe your solution in more depth.
How does it work? Why is this the best solution?
Does it solve a problem that affects multiple users or is this an edge case for your setup?
-->

This endpoint currently just maps all library items `toOldJSONMinified()` regardless of the `minified` parameter sent in the request. This change checks the parameter and maps to either the minified, or the `toOldJSONExpanded()`.

## How have you tested this?

<!-- Please describe in detail with reproducible steps how you tested your changes. -->

Tested by running locally through API test client with both response prior, and response after (with `minified=1`) and then with `minified=0`.

Also verified that change didn't affect web / mobile clients

## Screenshots

<!-- If your PR includes any changes to the web client, please include screenshots or a short video from before and after your changes. -->

n/a

r0adkll committed:
Adding support for 'minified' parameter when fetching a library's items

r0adkll committed:
Made fix less verbose

r0adkll commented:
@advplyr Hello 👋 Would it be possible to get a review of this PR?

Vito0912 commented:
Just want to add that this could potentially break clients that currently use minified, because they think it's working correctly and then miss specific values from response.
This should be announced at least one version before imo

advplyr commented:
What data specifically are you needing that is not provided in minified? We will be improving the API at some point and I put in this "minified" concept in the API when it was a small project. Some users have hundreds of thousands of books and books with thousands of audio files, etc so fully expanded and fetching all items isn't practical in a real application

jbhul commented:
> Just want to add that this could potentially break clients that currently use minified, because they think it's working correctly and then miss specific values from response. This should be announced at least one version before imo

Clients that currently omit minified=1 as a URL param could definitely break if they are using content present in minified but not in the regular output (in the minified - regular - extended paradigm). If this fix were to be implemented a way to prevent downstream issues would be to default to minified. Obviously this wouldn't be consistent with other endpoints but just a thought.

---

The way this is discussed in the docs is "regular" is the default but in reality the default is minified presently and regular doesn't work (even if minified=0 is passed).

I agree with @advplyr that such a large output may not be needed and could present issues if a user has a very large library with this much information being sent in a single call. I do think that the API docs should be updated to include that minified is the default and only output option. I'm also guessing the user who did the PR might want/need the full output and adding an option, even with a warning that this may not be performant or require pagination, might be a good option too.

Vito0912 commented:
The API documentation is currently not being maintained, is outdated, and more or less end-of-life.

It doesn’t make much sense to make major changes to it right now, since the backend is being overhauled anyway.

I don’t want to presume to decide whether it should be updated for this specific case, but updating the docs might send the wrong signal - it would suggest the documentation is current, which it’s not. Many endpoints are missing from the site, and some of the response and request information is outdated (as in missing, to much or just wrong). 

---

But the question regarding why one would need a non minified output outside of one item still stands. What are the reasons you would want such an output?

jbhul commented:
@Vito0912 Ah gotcha, didn't know that about the API docs. The docs themselves don't mention it and just say WIP which communicates the inverse of EOL lol.

In any case if they are EOL I'd think no reason to update to correct or notate, since it's one of many issues and there's no real reason to address this one while ignoring others.

---

I am not the original person who requested so I'd defer @deanishe from the [original issue](https://github.com/advplyr/audiobookshelf/issues/2123) or the PR creator.

Vito0912 commented:
> Ah gotcha, didn't know that about the API docs. The docs themselves don't mention it and just say WIP which communicates the inverse of EOL lol.
I see. Yeah, you can go to the Repo and actually see a PR that removes minfied from the docs from 2024. Maybe considerable to add a notice @advplyr ? WIP indeed is a bit misleading.

r0adkll commented:
> But the question regarding why one would need a non minified output outside of one item still stands. What are the reasons you would want such an output?

@Vito0912 @advplyr My need in getting the full output was to provide a smoother UX in the client that I am building by having the more detailed information stored by the time a user opens a detail page for any given title. Rather, than hitting the detail page with minified information, then it updating when the network request finishes.  

The original driver behind this PR was a bug in my assumption that data would come back as expanded and accidentally wiping out the relational (i.e. expanded) data related to a given item. I've since fixed this assumption and the experience works as expected sans waiting on a network load.

I was unaware of 
> The API documentation is currently not being maintained, is outdated, and more or less end-of-life.

at the time of building this and the `minified` field still being part of the documentation and sought to restore the functionality here.

---

> Just want to add that this could potentially break clients that currently use minified, because they think it's working correctly and then miss specific values from response.

> Clients that currently omit minified=1 as a URL param could definitely break if they are using content present in minified but not in the regular output

This is a great callout / concern and given the existing usage and assumptions around this endpoint the better strategy would be to some sort of opt-in parameter for this (an `expanded = 1` maybe?) since `minified` isn't really a part of this request anyway. 

---

> Some users have hundreds of thousands of books and books with thousands of audio files, etc so fully expanding all items isn't practical in a real application

I would say that this is also true of the minified response as well given that level of volume. I believe that the expanded format has a place here especially with pagination, but I do agree that using the `minified` parameter (in a way that is default expanded) is not the best path for this change. If you are agreeable to enabling expanded support on this endpoint I can update this PR to instead use `expanded` as an opt-in parameter ensuring that we don't disrupt existing API usage.

r0adkll committed:
Made 'expanded' details an explicit parameter

r0adkll commented:
Ah sorry for the delay on this but I have updated this PR to require an explicit `expanded = 1` request parameter making the return of expanded information an opt-in experience so as to not break any existing implementations that don't send a `minified = 1` 

@advplyr @Vito0912 If I could get your input once more as I believe I have addressed initial concerns and several others have chimed in on the use cases of returning expanded information in this endpoint.
